File: Kalorie/Kalorie/Core/UseCases/UpdateMyCreatedMealUseCase.swift
//
//  UpdateMyCreatedMealUseCase.swift
//  Kalorie
//
//  Created by Josef Antoni on 20.08.2026.
//

import Foundation

protocol UpdateMyCreatedMealUseCaseProtocol {
    func callAsFunction(_ meal: MyCreatedMealDomain) async throws
}

struct UpdateMyCreatedMealUseCase: UpdateMyCreatedMealUseCaseProtocol {

    // MARK: - Properties

    private let dataProvider: any FirestoreDataProviderProtocol
    private let authProvider: any AuthProviderProtocol

    // MARK: - Init

    init(dataProvider: any FirestoreDataProviderProtocol, authProvider: any AuthProviderProtocol) {
        self.dataProvider = dataProvider
        self.authProvider = authProvider
    }

    // MARK: - Functions

    func callAsFunction(_ meal: MyCreatedMealDomain) async throws {
        guard let userId = authProvider.userId else { throw AuthError.notAuthenticated }
        if let error = MyCreatedMealValidation.validate(name: meal.name, ingredients: meal.ingredients) { throw error }
        let updated = MyCreatedMealDomain(
            id: meal.id,
            name: meal.name.trimmingCharacters(in: .whitespacesAndNewlines),
            ingredients: meal.ingredients,
            createdAt: meal.createdAt,
            updatedAt: .now
        )
        let dto = MyCreatedMealDTO(meal: updated)
        try await dataProvider.setAsync(dto, id: updated.id, in: Constants.Firestore.myCreatedMeals(userId: userId))
    }
}

#if DEBUG
struct UpdateMyCreatedMealUseCaseFake: UpdateMyCreatedMealUseCaseProtocol {

    // MARK: - Properties

    var shouldThrow = false

    // MARK: - Functions

    func callAsFunction(_ meal: MyCreatedMealDomain) async throws {
        if shouldThrow { throw URLError(.unknown) }
    }
}
#endif

File: Android/app/src/main/java/antoni/kalorie/core/usecases/UpdateMyCreatedMealUseCase.kt
package antoni.kalorie.core.usecases

import antoni.kalorie.core.auth.AuthError
import antoni.kalorie.core.auth.AuthProviderProtocol
import antoni.kalorie.core.models.MyCreatedMealDomain
import antoni.kalorie.core.models.MyCreatedMealValidation
import antoni.kalorie.core.networking.FirestoreDataProviderProtocol
import antoni.kalorie.core.networking.MyCreatedMealDTO
import antoni.kalorie.core.networking.setAsync
import antoni.kalorie.core.utils.Constants
import java.time.Instant

interface UpdateMyCreatedMealUseCaseProtocol {
    suspend operator fun invoke(meal: MyCreatedMealDomain)
}

class UpdateMyCreatedMealUseCase(
    private val dataProvider: FirestoreDataProviderProtocol,
    private val authProvider: AuthProviderProtocol,
) : UpdateMyCreatedMealUseCaseProtocol {

    // MARK: - Functions

    override suspend fun invoke(meal: MyCreatedMealDomain) {
        val userId = authProvider.userId ?: throw AuthError.NotAuthenticated
        MyCreatedMealValidation.validate(name = meal.name, ingredients = meal.ingredients, portions = meal.portions)?.let { throw it }
        val updated = meal.copy(name = meal.name.trim(), updatedAt = Instant.now())
        dataProvider.setAsync(MyCreatedMealDTO(updated), id = updated.id, inCollection = Constants.Firestore.myCreatedMeals(userId))
    }
}
